modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -202,21 +215,85 @@ public CompletableFuture<Void> finish(
-                // TODO: IGNITE-20033 TestOnly code, let's consider using Txn state map instead of states.
-                .thenRun(() -> changeState(txId, PENDING, commit ? COMMITED : ABORTED));
+        Function<Void, CompletableFuture<Void>> clo = ignored -> {
+            HybridTimestamp commitTimestamp = commit ? clock.now() : null;
+
+            TxFinishReplicaRequest req = FACTORY.txFinishReplicaRequest()
+                    .txId(txId)
+                    .timestampLong(clock.nowLong())
+                    .groupId(commitPartition)
+                    .groups(groups)
+                    .commit(commit)
+                    .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
+                    .term(term)
+                    .build();
+
+            return replicaService.invoke(recipientNode, req)
+                    // TODO: IGNITE-20033 TestOnly code, let's consider using Txn state map instead of states.
+                    .thenRun(() -> changeState(txId, PENDING, commit ? COMMITED : ABORTED));
+        };
+
+        if (!commit) {
+            AtomicReference<CompletableFuture<Void>> ref = new AtomicReference<>();
+            TxContext tuple = txCtxMap.compute(txId, (uuid, tuple0) -> {
+                if (tuple0 == null) {
+                    tuple0 = new TxContext(); // No writes enlisted.
+                }
+
+                if (tuple0.finishFut == null) {
+                    tuple0.finishFut = new CompletableFuture<>();
+                    ref.set(tuple0.finishFut);
+                }
+
+                return tuple0;
+            });
+
+            if (ref.get() != null) { // This is aborting thread.
+                clo.apply(null).handle((ignored, err) -> {
+                    if (err == null) {
+                        tuple.finishFut.complete(null);
+                    } else {
+                        tuple.finishFut.completeExceptionally(err);
+                    }
+                    return null;
+                });
+            }
+
+            return tuple.finishFut;
+        }
+
+        // Wait for commit acks first, then proceed with the finish request.
+        AtomicReference<CompletableFuture<Void>> ref = new AtomicReference<>();
+        TxContext tuple = txCtxMap.compute(txId, (uuid, tuple0) -> {
+            if (tuple0 == null) {
+                tuple0 = new TxContext(); // No writes enlisted.
+            }
+
+            if (tuple0.finishFut == null) {
+                tuple0.finishFut = new CompletableFuture<>();
+                ref.set(tuple0.finishFut);
+            }
+
+            return tuple0;
+        });
+
+        if (ref.get() != null) { // This is committing thread.
+            // All inflights have been completed before the finish.
+            if (tuple.inflights == 0) {
+                tuple.waitRepFut.complete(null);
+            }
+
+            return tuple.waitRepFut.thenCompose(clo).handle((ignored, err) -> {
```

Comment:
In order to be on the same page. We do wait for all in-flights to complete even in case of rollback, right?

Reply:
No we do not.